BUG: BooleanArray.value_counts dropna

TomAugspurger opened the pull request:
Closes https://github.com/pandas-dev/pandas/issues/30685

TomAugspurger committed:
BUG: BooleanArray.value_counts dropna

Closes https://github.com/pandas-dev/pandas/issues/30685

jbrockmendel commented:
LGTM pending green

jreback commented on pandas/tests/arrays/test_boolean.py:
```diff
@@ -856,3 +856,10 @@ def test_arrow_roundtrip():
     tm.assert_frame_equal(result, df)
+
+
+def test_value_counts_na():
```
is the result an object dtyped Series when dropna=True? (add a test as well)

jorisvandenbossche replied:
In neither cases (dropna True of False) is the result an object dtype series, it is always integer (it are just counts).

That said, should the result here rather be a nullable integer type? Not that there are nulls here, but in the light of "trying to return nullable types as much as possible from operations involving nullable types".

jreback replied:
hmm, yeah i think we should just move this to return a nullable integer (as this is new api). will promote consistency in the future.

TomAugspurger replied:
Hmm, OK, will update these.

TomAugspurger replied:
And so API breaking change for IntegerARrray.value_counts to return a nullalble int dtype too?

jorisvandenbossche replied:
I would consider doing that, yes.

jbrockmendel replied:
> That said, should the result here rather be a nullable integer type? Not that there are nulls here, but in the light of "trying to return nullable types as much as possible from operations involving nullable types".

I think the "trying to return nullable..." would suggest we would want to keep the dtype for the _index_ here, not the values.  It seems weird that the values would be anything other than np.int64

TomAugspurger replied:
This discussion was about the values, not the index. I think the idea is to keep operations on these nullable types in the nullable type arena.

jbrockmendel replied:
Right, and I'm saying it _doesnt_ make sense to keep the values nullable

TomAugspurger replied:
Can you expand on why you find it weird?

It's true that the result of a value_counts will always have no NAs, but returning a nullable int type prevents a reintroduction of NAs in subsequent operations from converting to float.

jbrockmendel replied:
> Can you expand on why you find it weird?

The motivation is to maintain consistency of "operations with nullable types return nullable types".  But making value_counts().values return IntNA breaks the consistency of "values_counts().values is always np.int64".  So it's a wash on "maintaining consistency".

Ideally we'd retain the dtype in the value_counts().index, and it seems like we're saying here "well we cant do that, so let's shoehorn the dtype into the values"

TomAugspurger replied:
> it seems like we're saying here "well we cant do that, so let's shoehorn the dtype into the values"

No, I don't think we're saying that. I think we're saying we find a nullable integer dtype to be more useful.

jbrockmendel replied:
Not a hill I want to die on.

jreback replied:
commented below

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into 30685-value_counts

TomAugspurger committed:
fixup

TomAugspurger commented:
OK, updated {String,Integer,Boolean}Array.value_counts to return a nullable int64 dtype.

I moved BooleanArray.value_counts and IntegerArray.value_counts to the base class.

TomAugspurger committed:
remove unused numpy_dtype

jorisvandenbossche commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -411,6 +411,25 @@ Use :meth:`arrays.IntegerArray.to_numpy` with an explicit ``na_value`` instead.
+.. ipython:: python
+
+   >>> pd.Series([2, 1, 1, None], dtype="Int64").value_counts().dtype
+   Int64Dtype()
```
Don't show the output (or >>> prompt), or otherwise also make it a code-block

TomAugspurger replied:
Whoops, thanks.

jorisvandenbossche reviewed (commented):
Minor comment, looks good for the rest

TomAugspurger committed:
doc fixup

jreback commented:
let me clarify my comments

i think the index should be nullable ints as that preserves the intent here
but agree the values should be int64 - this is a count after all

TomAugspurger commented:
Thanks for clarifying.

The index can't be nullable yet. @jorisvandenbossche thoughts?

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -411,6 +411,24 @@ Use :meth:`arrays.IntegerArray.to_numpy` with an explicit ``na_value`` instead.
    a.to_numpy(dtype="float", na_value=np.nan)
 
+**value_counts returns a nullable integer dtype**
+
```
actually, even though is a breaking change, it puts nullable in more general usage. so actually i think this is a good change.

jorisvandenbossche commented:
> Ideally we'd retain the dtype in the value_counts().index, and it seems like we're saying here "well we cant do that, so let's shoehorn the dtype into the values"

IMO, ideally we retain the nullable dtype in *both* the index and the values. But indeed, for index we can't do that yet, so for now it's only the values.

>  It seems weird that the values would be anything other than np.int64

We are still returning an int dtype. But in the "nullable dtype universe"-subsystem, the "int64" dtype is the nullable int64 dtype.

TomAugspurger commented:
So I think that's some +1s and a -0 (https://github.com/pandas-dev/pandas/pull/30824#discussion_r364862869). Merging in an hour or so unless @jbrockmendel objects.

jbrockmendel commented:
No objection here.

TomAugspurger commented:
Thanks.
